### DISCUSSION TITLE
Parameter `target` in CaptumExplainer

### DISCUSSION THREAD

**Discussion by CYQFWang** [pyg-team/pytorch_geometric#7812] [Ideas]
I find following codes in `torch_geometric\explain\algorithm\captum_explainer.py`
Line 157
```python
        if self.model_config.mode == ModelMode.regression:
            target = None
        else:
            target = target[index]

        attributions = self.attribution_method.attribute(
            inputs=inputs,
            target=target,
            additional_forward_args=add_forward_args,
            **self.kwargs,
        )
``` 
Since `target` are "Output indices (for classification cases, this is usually the target class)", I think `target` ought to be `None` when `self.model_config.mode ==ModelMode.binary_classification`?

**Comment by rusty1s** [comment]
@RBendias Do you know why we set `target` to `None` here?
@CYQFWang Do you mean regression rather than binary classification?

**Reply by RBendias** [reply]
@CYQFWang Do you mean why we didn't set the `target` to `None` for binary classification? 

@rusty1s In the regression case, we set the `target` to `None`, because there doesn't need to be a target if there is only one scalar output per example.

**Reply by CYQFWang** [reply]
Yes, that's my question.

**Reply by CYQFWang** [reply]
@RBendias I have another question about Line 153 in the same file: 
```python
        self.attribution_method = self.attribution_method(captum_model)
``` 
When I use the same explainer to explain the second data, it raises an error saying `self.attribution_method` is not callable.
Then I think we should only do this setting at the first time?
```python
        if callable(self.attribution_method): self.attribution_method = self.attribution_method(captum_model)
```

**Reply by RBendias** [reply]
@CYQFWang Thanks for pointing out these two issues. I will send a PR to fix them.

**Reply by RBendias** [reply]
@CYQFWang, going over this again, I noticed that the current solution is correct. Captum expects the target to be output indices. According to our PyG implementation, this is true for binary classification or multilabel classification but not for regression:

Internally the target is computed with:
```
def get_target(self, prediction: Tensor) -> Tensor:
          r"""Returns the target of the model from a given prediction.
  
          If the model mode is of type :obj:`"regression"`, the prediction is
          returned as it is.
          If the model mode is of type :obj:`"multiclass_classification"` or
          :obj:`"binary_classification"`, the prediction is returned as the
          predicted class label.
```

While we can use the target as it is as output indices, for regression we can't. However, Captum only requires us to set a target if there is more than one output class per sample, which is not the case for regression which is why we simply set target=None.

**Reply by CYQFWang** [reply]
@RBendias Captum uses this parameter in (https://github.com/pytorch/captum/blob/master/captum/_utils/common.py#L583)
```python
    return output[(slice(None), *target)]
``` 
The output usually is a scalar for binary classification and it will raise a IndexError saying "index 1 is out of bounds for dimension 1 with size 1" Therefore, we probably should distinguish output indices from class label and not use class label as the `target` parameter in classification.

**Reply by RBendias** [reply]
@CYQFWang We convert binary-classification to multi-classification internally within the CaptumModel: 

```
        # Convert binary classification to multi-class classification:
        if (self.model_config is not None
                and self.model_config.mode == ModelMode.binary_classification):
            assert self.model_config.return_type == ModelReturnType.probs
            x = x.view(-1, 1)
            x = torch.cat([1 - x, x], dim=-1)
 ```
 
 Therefore, this shouldn't fail. Do you get an error anyway?

**Reply by CYQFWang** [reply]
@RBendias I think Captum wants a scalar as output for IntegratedGradients and this is why we are asked to provide output indices when our model actually have multi-dimensional output. 
```python
        assert outputs[0].numel() == 1, (
            "Target not provided when necessary, cannot"
            " take gradient with respect to multiple outputs."
        )
``` 
(https://github.com/pytorch/captum/blob/master/captum/_utils/gradient.py#L113)
Therefore, the `target` parameter should only depends on the shape of our model's output decided in Line151 as 
```python
            captum_model = CaptumModel(model, mask_type, index)
``` 
What we should do probaly is make sure that this model's output is a scalar and provide `target` if the model gives us a n-dimensional vector.

**Reply by RBendias** [reply]
According to my understanding of the code, the model will only return a scalar in case of regression (here, we do not provide the target). Due to the transformation in `CaptumModel`, we will always have a multi-dimensional output for classification (both binary and multi-class), here, we provide `target`.

**Reply by CYQFWang** [reply]
@RBendias  I'm afraid `attribution_method` in `captum.attr` doesn't want any multi-dimensional output model base on `assert outputs[0].numel() == 1`?

**Reply by RBendias** [reply]
If there is multi-dimensional output captum uses the target to transform it to a scalar [here](https://github.com/pytorch/captum/blob/945c582cc0b08885c4e2bfecb020abdfac0122f3/captum/_utils/common.py#L501).

**Reply by CYQFWang** [reply]
Yes, I think it works after the converting you mentioned and the `target` in `attribution_method` should be distinguished from the `target` provided by users in `Explainer`.
